staging/src/k8s.io/client-go/tools/record/events_cache.go

```diff
@@ -66,6 +74,105 @@ func DefaultEventFilterFunc(event *v1.Event) bool {
 	return false
 }
 
+// EventSpamFilter is responsible for filtering spam events.
+type EventSpamFilter struct {
+	sync.RWMutex
+
+	// the cache that manages last synced state
+	cache *lru.Cache
+
+	// the maximum number of same events that must occur before we begin filtering.
+	maxEvents int
+
+	// the amount of time in seconds that must transpire between messages to not be considered potential spam
+	maxIntervalInSeconds int
+
+	// the amount of time in seconds that transpire for a spam event to be allowed to sync (i.e. to bump the count)
+	syncIntervalInSeconds int
+
+	// clock is used to allow for testing over a time interval
+	clock clock.Clock
+}
+
+// NewEventSpamFilter knows how to perform spam filtering of matching events.
+func NewEventSpamFilter(lruCacheSize, maxEvents, maxIntervalInSeconds, syncIntervalSeconds int, clock clock.Clock) *EventSpamFilter {
+	return &EventSpamFilter{
+		cache:                 lru.New(lruCacheSize),
+		maxEvents:             maxEvents,
+		maxIntervalInSeconds:  maxIntervalInSeconds,
+		syncIntervalInSeconds: syncIntervalSeconds,
+		clock: clock,
+	}
+}
+
+// spamRecord holds data used to perform spam filtering decisions.
+type spamRecord struct {
+	// the amount of times this event appeared
+	numEvents int
+	// the last time at which the event was recorded.
+	lastTimestamp metav1.Time
+	// the last time at which the event was synced (to bump the count)
+	lastSynced metav1.Time
+}
+
+// Filter checks if the same event has been seen multiple times
+// in a sample window.  If it was seen more than the allowed number of
+// times in the rolling window, it will only let an event through
+// every sinkInterval.
+func (f *EventSpamFilter) Filter(event *v1.Event) bool {
+	var record spamRecord
+
+	now := metav1.NewTime(f.clock.Now())
+
+	// event key is the full cache key for this event
+	eventKey := getEventKey(event)
+
+	// do we have a record of similar events in our cache?
+	f.Lock()
+	defer f.Unlock()
+	value, found := f.cache.Get(eventKey)
+	if found {
+		record = value.(spamRecord)
+	}
+
+	// Is the previous record too old? If so, make a fresh one. Note: if we didn't
+	// find a similar record, its lastTimestamp will be the zero value, so we
+	// create a new one in that case.
+	maxInterval := time.Duration(f.maxIntervalInSeconds) * time.Second
+	interval := now.Time.Sub(record.lastTimestamp.Time)
+	if interval > maxInterval {
+		// enough time has transpired, we create a new record
+		record = spamRecord{}
+	}
```

Comment:
This is somewhat confusing.  I'd move L138-146 inside the if block at L134 and only overwrite `record` if `interval <  maxInterval`.  That way we don't create a spamRecord on L123, lose it on L135, and make another new one on L145.